src/sudoku.py

```diff
@@ -30,17 +30,33 @@ def __init__(self, grid: Grid) -> None:
             self.__squares.append([self.__rows[x][y] for x in range(i, i+3) for y in range(j, j+3)])
         self._speculation_depth = 0
 
+    @staticmethod
+    def __leave_equal_alternatives(grid_view: list[list[Cell]]) -> bool:
+        is_any_cell_solved = False  # флаг что хоть одна клетка решена
+        for batch in grid_view:
+            list_of_alt_sets = [batch[j].alternatives for j in range(9)]    # список альтернатив в одной пачке
+            alternative_to_cell_indexes = {m: frozenset(n for n in range(9) if m in list_of_alt_sets[n]) for m in range(1, 10)}  # словарь цифра: список множества клеток, в которых она встречается
+            twin_cell_list = Counter(alternative_to_cell_indexes[m] for m in alternative_to_cell_indexes)  # счетчик одинаковых списков клеток
+            for cells in twin_cell_list:      # перебираем счетчик
+                if len(cells) == twin_cell_list[cells]:  # если длина списка клеток равна количеству цифр, которые встречаются в этих клетках
+                    exclude_set = {k for k in alternative_to_cell_indexes if alternative_to_cell_indexes[k] == cells}
+                    for cell in cells:  # из клеток в найденном множестве cells... 
+                        batch[cell].exclude(batch[cell].alternatives.difference(exclude_set)) # исключаем все, кроме exclude_set
```

Comment:
1. У тебя уже собраны в список альтернативы клеток, нет нужды заново вызывать метод `.alternatives`
```suggestion
                        batch[j].exclude(list_of_alt_sets[j].difference(exclude_set))
```
2. Лучше читается простая арифметика:

```suggestion
                        batch[j].exclude(list_of_alt_sets[j] - exclude_set) 
```
3. Кстати, название `exclude_set` тоже вводит в заблуждение. Лучше что-то вроде `common_alts`.